pkg/controller/clusterrelocate/clusterrelocate_controller.go

```diff
@@ -0,0 +1,616 @@
+	clusterRelocate := a.Object.(*hivev1.ClusterRelocate)
+
+	labelSelector, err := metav1.LabelSelectorAsSelector(&clusterRelocate.Spec.ClusterDeploymentSelector)
+	if err != nil {
+		r.logger.WithError(err).
+			WithField("clusterRelocate", clusterRelocate.Name).
+			Warn("cannot parse clusterdeployment selector")
+		return
+	}
+
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	if err := r.List(context.Background(), clusterDeployments); err != nil {
+		r.logger.WithError(err).
+			WithField("clusterRelocate", clusterRelocate.Name).
+			Log(controllerutils.LogLevel(err), "failed to list clusterdeployments")
+		return
+	}
+
+	for _, cd := range clusterDeployments.Items {
+		if cd.Annotations[constants.RelocatingAnnotation] != clusterRelocate.Name &&
+			!labelSelector.Matches(labels.Set(cd.Labels)) {
+			continue
+		}
+		requests = append(requests,
+			reconcile.Request{NamespacedName: types.NamespacedName{Name: cd.Name, Namespace: cd.Namespace}})
+	}
+
+	return
+}
+
+// ReconcileClusterRelocate is the reconciler for ClusterRelocate. It will sync on ClusterDeployment resources
+// and relocate those that match with a ClusterRelocate.
+type ReconcileClusterRelocate struct {
+	client.Client
+	scheme *runtime.Scheme
+	logger log.FieldLogger
+
+	// remoteClusterAPIClientBuilder is a function pointer to the function that gets a builder for building a client
+	// for the remote cluster's API server
+	remoteClusterAPIClientBuilder func(secret *corev1.Secret) remoteclient.Builder
+}
+
+// Reconcile relocates ClusterDeployments matching with a ClusterRelocate to another Hive instance.
+func (r *ReconcileClusterRelocate) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+	logger := r.logger.WithFields(log.Fields{
+		"controller":        controllerName,
+		"clusterDeployment": request.NamespacedName.String(),
+	})
+
+	// For logging, we need to see when the reconciliation loop starts and ends.
+	logger.Info("reconciling cluster deployment")
+	defer func() {
+		dur := time.Since(start)
+		hivemetrics.MetricControllerReconcileTime.WithLabelValues(controllerName).Observe(dur.Seconds())
+		logger.WithField("elapsed", dur).Info("reconcile complete")
+	}()
+
+	// Fetch the ClusterDeployment instance
+	cd := &hivev1.ClusterDeployment{}
+	err := r.Get(context.TODO(), request.NamespacedName, cd)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Debug("cluster deployment not found")
+			return reconcile.Result{}, nil
+		}
+		logger.WithError(err).Error("Error getting cluster deployment")
+		return reconcile.Result{}, err
+	}
+
+	if cd.DeletionTimestamp != nil {
+		logger.Debug("skipping deleted clusterdeployment")
+		return reconcile.Result{}, err
+	}
+
+	desiredRelocators, err := r.findMatchingRelocators(cd, logger)
+
+	if len(desiredRelocators) != 1 {
+		if err := r.stopRelocating(cd, true, logger); err != nil {
```

Comment:
Please log a warning with the desired relocators count so we have context as to why we're trying to stop relocating in the case where it errors.

Reply:
The `matchingRelocators` field has been added to the log entry.